### PR TITLE
Ensure we have a bound node for the unsafe block so that region analysis works correctly.

### LINKED ISSUE #4950: Extract Method refactoring fails to capture inputs/outputs when unsafe blocks are included
From @AArnott:

``` C#
using System;

class Program {
    static void Main(string[] args) {
        object value = args;

        // extract from here
        IntPtr p;
        unsafe
        {
            object t = value;
            p = IntPtr.Zero;
        }
        // to here

        Console.WriteLine(p);
    }
}
```

Open the C# console app in Dev14 and the Program.cs file.
Select lines 7 - 12 and execute the Extract Method command.

_Expected_
A new method that takes an object as a parameter and returns an IntPtr is created, and an invocation of that method replaces the selected lines that passes in "value" as the parameter and assigns the result to "IntPtr p". 

_Actual_
A method is created that takes no parameters and returns void. Two compile errors are created by the refactoring as a result.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #4950

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 1]
```diff
@@ -1474,6 +1474,13 @@ private IParameterInitializerOperation CreateBoundParameterEqualsValueOperation(
 
         private IBlockOperation CreateBoundBlockOperation(BoundBlock boundBlock)
         {
+            // The C# binder produces an extra block for checked, unchecked, and unsafe statements, but they have no semantics of their own
+            boundBlock = boundBlock switch
+            {
+                { Syntax: CheckedStatementSyntax _, Locals: { IsDefaultOrEmpty: true }, Statements: { Length: 1 } s } => (BoundBlock)s[0],
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1480]: >Syntax: CheckedStatementSyntax [](start = 18, length = 30)

I think we are trying to follow the principle that semantic meaning of bound nodes shouldn't depend on their associated syntax. This condition violates the principle.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1480]: >Locals [](start = 52, length = 6)

The block could also have ```LocalFunctions```.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1480]: >(BoundBlock) [](start = 121, length = 12)

The unconditional cast makes the code fragile, I believe.

**Comment by AlekseyTs** [comment]
@gafter 
> Ensure we have a bound node for the unsafe block so that region analysis works correctly.

Could you please elaborate why is that dummy block necessary for correct behavior of the region analysis. Do we have the same issue with parenthesized expressions missing from the bound tree? If we don't, could we treat unsafe/checked/unchecked statements in a similar manner?

**Comment by AlekseyTs** [comment]
I suspect we are failing to detect when we enter and leave the region. Could we simply dig through these statements when we determining the boundaries?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs [hunk 2]
```diff
@@ -11283,5 +11283,62 @@ string NewMethod()
...
+using System;
+
+class Program {
+    static void Main(string[] args)
+    {
+        object value = args;
+
+        [|
+        IntPtr p;
+        unsafe
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:11303]:
```suggestion
        {keyword}
```

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs [hunk 3]
```diff
@@ -11283,5 +11283,62 @@ string NewMethod()
...
+
+        IntPtr p = NewMethod(value);
+
+        Console.WriteLine(p);
+    }
+
+    private static IntPtr NewMethod(object value)
+    {
+        IntPtr p;
+        unsafe
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:11330]:
```suggestion
        {keyword}
```

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs [hunk 4]
```diff
@@ -11283,5 +11283,62 @@ string NewMethod()
...
+        {
+            object t = value;
+            p = IntPtr.Zero;
+        }
+
+        return p;
+    }
+}
+";
+                await TestExtractMethodAsync(code.Replace("unsafe", keyword), expected.Replace("unsafe", keyword));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:11340]:
```suggestion
                await TestExtractMethodAsync(code, expected);
```

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs [hunk 5]
```diff
@@ -11283,5 +11283,62 @@ string NewMethod()
 }";
             await TestExtractMethodAsync(code, expected);
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.ExtractMethod)]
+        [WorkItem(4950, "https://github.com/dotnet/roslyn/issues/4950")]
+        public async Task ExtractMethodInvolvingUnsafeBlock()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:11289]:
```suggestion
        [Theory, InlineData("unsafe", "checked", "unchecked")]
        [Trait(Traits.Feature, Traits.Features.ExtractMethod)]
        [WorkItem(4950, "https://github.com/dotnet/roslyn/issues/4950")]
        public async Task ExtractMethodInvolvingUnsafeBlock(string keyword)
```

File: src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs [hunk 6]
```diff
@@ -11283,5 +11283,62 @@ string NewMethod()
...
+        {
+            foreach (string keyword in new[] { "unsafe", "checked", "unchecked" })
+            {
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:11292]:
```suggestion
```

**Comment by gafter** [comment]
@AlekseyTs @CyrusNajmabadi I think I've responded to all of your comments. Do you have any others?

@333fred Would you please review this PR?

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 7]
```diff
@@ -1948,14 +1948,12 @@ protected void EnsureNullabilityAnalysisPerformedIfNecessary()
...
 
-            Debug.Assert(Root == bindableRoot);
-
             NullableWalker.SnapshotManager snapshotManager;
             var remappedSymbols = _parentRemappedSymbolsOpt;
             BoundNode boundRoot;
             Binder binder;
 
-            bind(bindableRoot, out binder, out boundRoot);
+            bind(Root, out binder, out boundRoot);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1956]: >Root [](start = 17, length = 4)

Is there a requirement to bind Root in our case? Is this going to have an observable difference for our case? This feels fragile,  GetBindableSyntaxNode is designed to give back the node that can be bound on its own, other consumers are likely to be negatively impacted. I would like to better understand the reason for the adjustments in this function. If it is just due to the Assert, I would prefer us to remove, or adjust the assert instead.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 8]
```diff
@@ -1948,14 +1948,12 @@ protected void EnsureNullabilityAnalysisPerformedIfNecessary()
 
             upgradeableLock.EnterWrite();
 
-            Debug.Assert(Root == bindableRoot);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1951]: Please test IOperation behavior for when nullable is enabled and the `unchecked`/`unsafe` block is the root. In particular, what the behavior of calling `SemanticModel.GetIOperation` on that `UnsafeStatementSyntax` is. The `Type` property of those nodes will be wrong unless you wait for [https://github.com/dotnet/roslyn/pull/42403](https://github.com/dotnet/roslyn/pull/42403) to merge and merge 16.8-p1 into your local code, but I'd expect other symbols like the method symbols of `IInvocationOperation`s or local symbols to reflect the correct nullability.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1951]: > In particular, what the behavior of calling SemanticModel.GetIOperation on that UnsafeStatementSyntax is.

I would expect to get nothing as for any node that disappears from the tree, parenthesized and such.

> The Type property of those nodes will be wrong unless you wait for [https://github.com/dotnet/roslyn/pull/42403](https://github.com/dotnet/roslyn/pull/42403) to merge and merge 16.8-p1 into your local code, but I'd expect other symbols like the method symbols of IInvocationOperations or local symbols to reflect the correct nullability.

These are statements, they do not have types, and I expect to get no IOperation node for them.


---
In reply to: [444547469](https://github.com/dotnet/roslyn/pull/45334#discussion_r444547469) [](ancestors = 444547469)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs (hunk 7 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:1956]: No, as the relevant semantics are already encoded in the binder.

---
In reply to: [444547175](https://github.com/dotnet/roslyn/pull/45334#discussion_r444547175) [](ancestors = 444547175)

**Comment by gafter** [comment]
@AlekseyTs I think I've responded to all of your comments. Do you have any others?

**Pull request merged**